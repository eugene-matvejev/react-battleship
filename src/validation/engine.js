const executeValidations = (item, c) => {
    const { validators, value } = item;

    if (!Array.isArray(validators)) {
        return true;
    }

    item.errors = [];

    let isValid = true;

    for (const fn of validators) {
        const r = fn(value, c);
        /**
         * if result is a string, then it is considered as error, otherwise it considered as valid
         */
        if (typeof r === 'string') {
            isValid = false;

            item.errors.push(r);
        }
    }

    return isValid;
}

/**
 * @param {Array} c
 * @param {Array} enforceValidationOn
 *
 * example of enforceValidationOn
 *  [[sectionId, itemId]] - will validate only item in specific section
 *  [[sectionId, undefined]] - will validate entire section, if item is not defined
 */
export const validationEngine = (c, enforceValidationOn) => {
    let isValid = true;

    if (Array.isArray(enforceValidationOn)) {
        for (const [sectionId, itemId] of enforceValidationOn) {
            if (undefined === itemId) {
                for (const item of c[sectionId].items) {
                    if (!executeValidations(item, c)) {
                        isValid = false;
                    }
                }

                continue;
            }

            if (!executeValidations(c[sectionId].items[itemId], c)) {
                isValid = false;
            }
        }

        return isValid;
    }

    for (const { items } of c) {
        for (const item of items) {
            if (!executeValidations(item, c)) {
                isValid = false;
            }
        }
    }

    return isValid;
}
