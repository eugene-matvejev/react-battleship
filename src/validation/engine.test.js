import { validationEngine } from './engine';

describe('validation engine', () => {
    const c = [
        {
            items: [
            ],
        },
        {
            items: [
            ],
            items: [
            ],
        },
    ];

    describe('[default mode]', () => {
        it(`should return TRUE when validators are NOT defined, expected NO error messages in snapshot`, () => {
            c[0].items[0] = {}
            c[1].items[0] = {};
            c[1].items[1] = {};

            const v = validationEngine(c);

            expect(v).toBe(true);
            expect(c).toMatchSnapshot();
        });

        it(`should return TRUE when validators are empty, expected NO error messages in snapshot`, () => {
            c[0].items[0] = {
                validators: [],
            };
            c[1].items[0] = {};
            c[1].items[1] = {};

            const v = validationEngine(c);

            expect(v).toBe(true);
            expect(c).toMatchSnapshot();
        });

        it(`should return TRUE when all defined validators PASS, expected NO error messages in snapshot`, () => {
            c[0].items[0] = {
                validators: [
                    () => true,
                ],
            };
            c[1].items[0] = {};
            c[1].items[1] = {};


            const v = validationEngine(c);

            expect(v).toBe(true);
            expect(c).toMatchSnapshot();
        });

        it(`should return FALSE when all defined validators FAIL, expected error messages in snapshot`, () => {
            c[0].items[0] = {
                validators: [
                    () => '{{error message}}',
                ],
            };
            c[1].items[0] = {};
            c[1].items[1] = {};

            const v = validationEngine(c);

            expect(v).toBe(false);
            expect(c).toMatchSnapshot();
        });

        it(`should return FALSE when at least one of defined validators FAIL, expected error messages in snapshot`, () => {
            c[0].items[0] = {
                validators: [
                    () => '{{error message}}',
                    () => true,
                ],
            };
            c[1].items[0] = {};
            c[1].items[1] = {};

            const v = validationEngine(c);

            expect(v).toBe(false);
            expect(c).toMatchSnapshot();
        });
    });

    describe('[enforced scope mode]', () => {
        describe('[scope limited to entire section]', () => {
            it(`should return TRUE when all validators in a scope PASS`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    value: '{{item-1-0}}',
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(),
                    ],
                };

                expect(validationEngine(c, [[1]])).toBe(true);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).toBeCalledWith('{{item-1-0}}', c);
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });

            it(`should return FALSE when all validators in a scope FAIL`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    value: '{{item-1-0}}',
                    validators: [
                        jest.fn(() => 'error'),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(() => 'error'),
                    ],
                };

                expect(validationEngine(c, [[1]])).toBe(false);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).toBeCalledWith('{{item-1-0}}', c);
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });

            it(`should return FALSE when at least one validator in a scope FAIL`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    value: '{{item-1-0}}',
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(() => 'error'),
                    ],
                };

                expect(validationEngine(c, [[1]])).toBe(false);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).toBeCalledWith('{{item-1-0}}', c);
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });
        });

        describe('[scope limited to specific item]', () => {
            it(`should return TRUE when all validators in a scope PASS`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(),
                    ],
                };

                expect(validationEngine(c, [[1, 1]])).toBe(true);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });

            it(`should return FALSE when all validators in a scope FAIL`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(() => 'error'),
                        jest.fn(() => 'error'),
                    ],
                };

                expect(validationEngine(c, [[1, 1]])).toBe(false);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);
                expect(c[1].items[1].validators[1]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });

            it(`should return FALSE when at least one validator in a scope FAIL`, () => {
                c[0].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[0] = {
                    validators: [
                        jest.fn(),
                    ],
                };
                c[1].items[1] = {
                    value: '{{item-1-1}}',
                    validators: [
                        jest.fn(),
                        jest.fn(() => 'error'),
                    ],
                };

                expect(validationEngine(c, [[1, 1]])).toBe(false);

                expect(c[0].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[0].validators[0]).not.toBeCalled();
                expect(c[1].items[1].validators[0]).toBeCalledWith('{{item-1-1}}', c);
                expect(c[1].items[1].validators[1]).toBeCalledWith('{{item-1-1}}', c);

                expect(c).toMatchSnapshot();
            });
        });
    });
});
