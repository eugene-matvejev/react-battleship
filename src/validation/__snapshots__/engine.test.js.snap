// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation engine [default mode] should return FALSE when all defined validators FAIL, expected error messages in snapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "errors": Array [
          "{{error message}}",
        ],
        "validators": Array [
          [Function],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {},
      Object {},
    ],
  },
]
`;

exports[`validation engine [default mode] should return FALSE when at least one of defined validators FAIL, expected error messages in snapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "errors": Array [
          "{{error message}}",
        ],
        "validators": Array [
          [Function],
          [Function],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {},
      Object {},
    ],
  },
]
`;

exports[`validation engine [default mode] should return TRUE when all defined validators PASS, expected NO error messages in snapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "errors": Array [],
        "validators": Array [
          [Function],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {},
      Object {},
    ],
  },
]
`;

exports[`validation engine [default mode] should return TRUE when validators are NOT defined, expected NO error messages in snapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {},
    ],
  },
  Object {
    "items": Array [
      Object {},
      Object {},
    ],
  },
]
`;

exports[`validation engine [default mode] should return TRUE when validators are empty, expected NO error messages in snapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "errors": Array [],
        "validators": Array [],
      },
    ],
  },
  Object {
    "items": Array [
      Object {},
      Object {},
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to entire section] should return FALSE when all validators in a scope FAIL 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "errors": Array [
          "error",
        ],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-0}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
        ],
        "value": "{{item-1-0}}",
      },
      Object {
        "errors": Array [
          "error",
        ],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to entire section] should return FALSE when at least one validator in a scope FAIL 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "errors": Array [],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-0}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        ],
        "value": "{{item-1-0}}",
      },
      Object {
        "errors": Array [
          "error",
        ],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to entire section] should return TRUE when all validators in a scope PASS 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "errors": Array [],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-0}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        ],
        "value": "{{item-1-0}}",
      },
      Object {
        "errors": Array [],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to specific item] should return FALSE when all validators in a scope FAIL 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
      Object {
        "errors": Array [
          "error",
          "error",
        ],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to specific item] should return FALSE when at least one validator in a scope FAIL 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
      Object {
        "errors": Array [
          "error",
        ],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": "error",
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;

exports[`validation engine [enforced scope mode] [scope limited to specific item] should return TRUE when all validators in a scope PASS 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "validators": Array [
          [MockFunction],
        ],
      },
      Object {
        "errors": Array [],
        "validators": Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "{{item-1-1}}",
                [Circular],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        ],
        "value": "{{item-1-1}}",
      },
    ],
  },
]
`;
